resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/BaseSubmissionStep.scala

```diff
@@ -0,0 +1,136 @@
+import org.apache.spark.deploy.kubernetes.config._
+import org.apache.spark.deploy.kubernetes.constants._
+
+/**
+ * Represents the initial setup required for the driver.
+ */
+private[spark] class BaseSubmissionStep(
+      kubernetesAppId: String,
+      kubernetesResourceNamePrefix: String,
+      driverLabels: Map[String, String],
+      dockerImagePullPolicy: String,
+      appName: String,
+      mainClass: String,
+      appArgs: Array[String],
+      submissionSparkConf: SparkConf)
+    extends KubernetesSubmissionStep {
+
+  private val kubernetesDriverPodName = submissionSparkConf.get(KUBERNETES_DRIVER_POD_NAME)
+    .getOrElse(s"$kubernetesResourceNamePrefix-driver")
+  private val driverExtraClasspath = submissionSparkConf.get(
+      org.apache.spark.internal.config.DRIVER_CLASS_PATH)
+  // CPU settings
+  private val driverCpuCores = submissionSparkConf.getOption("spark.driver.cores").getOrElse("1")
+  private val driverLimitCores = submissionSparkConf.getOption(KUBERNETES_DRIVER_LIMIT_CORES.key)
+
+  // Memory settings
+  private val driverMemoryMb = submissionSparkConf.get(
+        org.apache.spark.internal.config.DRIVER_MEMORY)
+  private val memoryOverheadMb = submissionSparkConf
+      .get(KUBERNETES_DRIVER_MEMORY_OVERHEAD)
+      .getOrElse(math.max((MEMORY_OVERHEAD_FACTOR * driverMemoryMb).toInt,
+          MEMORY_OVERHEAD_MIN))
+  private val driverContainerMemoryWithOverhead = driverMemoryMb + memoryOverheadMb
+  private val driverDockerImage = submissionSparkConf.get(DRIVER_DOCKER_IMAGE)
+
+  override def prepareSubmission(
+      driverSpec: KubernetesDriverSpec): KubernetesDriverSpec = {
+    val driverExtraClasspathEnv = driverExtraClasspath.map { classPath =>
+      new EnvVarBuilder()
+        .withName(ENV_SUBMIT_EXTRA_CLASSPATH)
+        .withValue(classPath)
+        .build()
+    }
+    val driverCpuQuantity = new QuantityBuilder(false)
+      .withAmount(driverCpuCores)
+      .build()
+    val driverMemoryQuantity = new QuantityBuilder(false)
+      .withAmount(s"${driverMemoryMb}M")
+      .build()
+    val driverMemoryLimitQuantity = new QuantityBuilder(false)
+      .withAmount(s"${driverContainerMemoryWithOverhead}M")
+      .build()
+    val maybeCpuLimitQuantity = driverLimitCores.map { limitCores =>
+      ("cpu", new QuantityBuilder(false).withAmount(limitCores).build())
+    }
+    val driverContainer = new ContainerBuilder(driverSpec.driverContainer)
+      .withName(DRIVER_CONTAINER_NAME)
+      .withImage(driverDockerImage)
+      .withImagePullPolicy(dockerImagePullPolicy)
+      .addToEnv(driverExtraClasspathEnv.toSeq: _*)
+      .addNewEnv()
+        .withName(ENV_DRIVER_MEMORY)
+        .withValue(driverContainerMemoryWithOverhead + "m")
+        .endEnv()
+      .addNewEnv()
+        .withName(ENV_DRIVER_MAIN_CLASS)
+        .withValue(mainClass)
+        .endEnv()
+      .addNewEnv()
+        .withName(ENV_DRIVER_ARGS)
+        .withValue(appArgs.mkString(" "))
+        .endEnv()
```

Comment:
Check most recent push that I made which resolves this and previous comments about argument issues